services/arrow-flight/src/main/java/com/dremio/service/flight/impl/FlightWorkManager.java

```diff
@@ -156,6 +162,80 @@ public void getCatalogs(FlightProducer.ServerStreamListener listener, BufferAllo
       listener.putNext();
       listener.completed();
     }
+    workerProvider.get().submitWork(runExternalId, userSession, responseHandler, userRequest, TerminationListenerRegistry.NOOP);
+  }
+
+  /**
+   * Submits a GET_TABLES job to a worker and sends the response to given ServerStreamListener.
+   *
+   * @param listener    ServerStreamListener listening to the job result.
+   * @param allocator   BufferAllocator used to allocate the response VectorSchemaRoot.
+   * @param userSession The session for the user which made the request.
+   */
+  public void runGetTables(FlightSql.CommandGetTables commandGetTables,
+                           FlightProducer.ServerStreamListener listener,
+                           Supplier<Boolean> isRequestCancelled,
+                           BufferAllocator allocator,
+                           UserSession userSession) {
+    final UserBitShared.ExternalId runExternalId = ExternalIdHelper.generateExternalId();
+    final UserProtos.GetTablesReq.Builder builder = UserProtos.GetTablesReq.newBuilder();
+
+    if (commandGetTables.hasSchemaFilterPattern()) {
+      builder.setSchemaNameFilter(UserProtos.LikeFilter.newBuilder()
+        .setPattern(commandGetTables.getSchemaFilterPattern().getValue()).build());
+    }
+
+    if (commandGetTables.hasTableNameFilterPattern()) {
+      builder.setTableNameFilter(UserProtos.LikeFilter.newBuilder()
+        .setPattern(commandGetTables.getTableNameFilterPattern().getValue()).build());
+    }
+
+    if (commandGetTables.hasCatalog()) {
+      builder.setCatalogNameFilter(UserProtos.LikeFilter.newBuilder()
+        .setPattern(commandGetTables.getCatalog().getValue()).build());
+    }
+
+    if (!commandGetTables.getTableTypesList().isEmpty()) {
+      builder.addAllTableTypeFilter(commandGetTables.getTableTypesList());
+    }
+
+    final UserRequest userRequest =
+      new UserRequest(UserProtos.RpcType.GET_TABLES, builder.build());
+
+    final CancellableUserResponseHandler<UserProtos.GetTablesResp> responseHandler =
+      new CancellableUserResponseHandler<>(runExternalId, userSession, workerProvider, isRequestCancelled,
+        UserProtos.GetTablesResp.class);
+
+    workerProvider.get().submitWork(runExternalId, userSession, responseHandler, userRequest, TerminationListenerRegistry.NOOP);
+
+    final UserProtos.GetTablesResp getTablesResp = responseHandler.get();
+
+    try (VectorSchemaRoot vectorSchemaRoot = VectorSchemaRoot.create(FlightSqlProducer.Schemas.GET_TABLES_SCHEMA,
```

Comment:
See comment regarding refactoring to reduce duplication on the CommandGetSchemas PR.

Reply:
We've tried to find a way to refactor this, but each schema has its different values making it difficult. Do you have any idea?

Reply:
That's fine then.